walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,114 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // If duration is longer than a day, no option for this case.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // Get general available time ranges.
+    Collection<String> attendees = request.getAttendees();
+    List<TimeRange> unavailableTimeWithOverlap = getAttendeesUnavailableTime(events, attendees);
+    List<TimeRange> unavailableTime = getNoOverlapTime(unavailableTimeWithOverlap);
+    List<TimeRange> availableTime = getAvailableTime(unavailableTime, request.getDuration());
+    return availableTime;
+  }
+
+  /** Get a collection of all attendees' unavailable time ranges (due to other events). */
+  private List<TimeRange> getAttendeesUnavailableTime(
+      Collection<Event> events, Collection<String> meetingAttendees) {
+    List<TimeRange> unavailableTime = new ArrayList<>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      // If any attendee in the meeting joins that event as well, the time range of the event is
+      // added to the unavailableTime collection.
+      if (!Collections.disjoint(meetingAttendees, eventAttendees)) {
+        unavailableTime.add(event.getWhen());
+      }
+    }
+    return unavailableTime;
+  }
+
+  /** Return a collection of time ranges in the time order without any overlap among each other. */
+  private List<TimeRange> getNoOverlapTime(List<TimeRange> unavailableTime) {
+    List<TimeRange> noOverlapTime = new ArrayList<>();
+    for (TimeRange newTime : unavailableTime) {
+      if (noOverlapTime.isEmpty()) {
+        noOverlapTime.add(newTime);
+        continue;
+      }
+      // Compare and combine the new time with any existing time range in collection in order.
+      List<TimeRange> localNoOverlapTime = new ArrayList<>();
```

Comment:
It may be helpful to write as a comment the loop invariant to help the reader understand the algorithm.

Reply:
Sure. I have just added new comment for "noOverlapTime" list (including some explanations for "localNoOverlapTime") for better understanding.